### PR TITLE
chore: bump tonic to v0.12 & etcd-client to v0.14

### PR DESCRIPTION
**Pull request by BugenZhao**

### CODE REVIEW

**Review by wangrunji0408 — APPROVED** [review]
Thank you 🥹

**Comment by wangrunji0408** [comment]
Could you help fix the clippy issue? 

https://github.com/madsim-rs/madsim/actions/runs/10177920150/job/28150455207?pr=222#step:6:539 🥵

File: Cargo.toml [hunk 1]
```diff
@@ -14,3 +14,9 @@ resolver = "2"
 
 [patch.crates-io]
 getrandom = { git = "https://github.com/madsim-rs/getrandom.git", rev = "6c9d9e9" }
+
+[workspace.lints.rust]
+unexpected_cfgs = { level = "warn", check-cfg = [
+    'cfg(madsim)',
+    'cfg(tokio_unstable)',
+] }
```
Comment (BugenZhao) [inline Cargo.toml:22]: New check introduced in rustc v1.80

File: madsim-tonic-build/src/prost.rs [hunk 2]
```diff
@@ -59,6 +60,134 @@ pub fn compile_protos(proto: impl AsRef<Path>) -> io::Result<()> {
     Ok(())
 }
 
+/// Non-path Rust types allowed for request/response types.
+const NON_PATH_TYPE_ALLOWLIST: &[&str] = &["()"];
```
Comment (BugenZhao) [inline madsim-tonic-build/src/prost.rs:64]: Copied from tonic v0.12

**Comment by BugenZhao** [comment]
@wangrunji0408 The PR is ready to merge. 🥰

**Comment by wangrunji0408** [comment]
published new versions to crates.io

**Pull request merged**